### PR TITLE
BIP-draft:  2mb blocksize step

### PR DESCRIPTION
**Pull request by jgarzik**
Folks had requested a variant of BIP 102 that includes a small step, to avoid the moral hazard "revisit this question" issue.

This has similar properties to BIP 102 - it is not intended to handle exponential growth, nor be a for-all-time solution. It is a short term adjustment, buying time for longer term thought, modeling and experimentation with a dynamic upper limit.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -3055,6 +3063,11 @@ bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& sta
         return state.Invalid(error("%s : rejected nVersion=3 block", __func__),
                              REJECT_OBSOLETE, "bad-version");
 
+    // Reject block.nVersion=4 blocks when 95% (75% on testnet) of the network has upgraded:
+    if (block.nVersion < 5 && IsSuperMajority(5, pindexPrev, consensusParams.nMajorityRejectBlockOutdated, consensusParams))
+        return state.Invalid(error("%s : rejected nVersion=3 block", __func__),
```
Comment (btcdrak) [inline src/main.cpp]: `nVersion=4`

**Comment by petertodd** [comment]
Did you get a BIP # allocation?

Don't allocate numbers yourself.

**Comment by jameshilliard** [comment]
I think we need an upper ceiling on any proposal we adopt, I think 8MB may be appropriate for the time being although I would prefer BIP102 in general since it seems safer.

**Comment by ChainQuery** [comment]
utACK

**Comment by omarabid** [comment]
@GamerSg Regarding 3. At 1MB/year increase, it'd take 10 years to reach 12MB block size. 10 years ago the iPhone didn't exist. 12MB + SW would mean around 5million tx/day. That's not enough even for settlement if Bitcoin was to handle even a small fraction of todays financial system.

**Comment by jgarzik** [comment]
This has similar properties to BIP 102 - it is not intended to handle exponential growth, nor be a for-all-time solution.  It is a short term adjustment, buying time for longer term thought, modeling and experimentation with a dynamic upper limit.

**Comment by Zaromet** [comment]
Sorry new at this so I have no clue how to add net.h and add increase in line 47. P2P masage... Limited to 2MB.

**Comment by petertodd** [comment]
If its a short term adjustment, why the perpetual growth?

Anyway, the tech fundamentally has O(n²) scaling, so no amount of dynamic fiddling is going to give us long term growth that's significant. We need fundamental changes for that.

**Comment by omarabid** [comment]
@jgarzik I have the same opinion as @petertodd. If it's not a perfect solution, then it doesn't need to "slowly" increase the block size limit. I think a fast 2-4-8 pump is more appropriate until a solid solution is designed.

In the last few months, the number of transactions jumped from around 100k per day to 200k. That's double in a few months. What if the number of transactions goes to 0.5million/day once we activate the 2Mb cap, we'll then have to revisit this temporary fix again.

**Comment by petertodd** [comment]
To be clear, "revisiting" doesn't mean the limit can or will be changed again. This is a security parameter and we can't simply raise it at will, and may even be forced to decrease.

Any increase must come with stakeholders signing onto the concept that this is a likely one-time increase. If not theres no reason for us to release code with an increase - let a competing team do that.

**Comment by hellobitcoins** [comment]
ACK

**Comment by ydtr** [comment]
ACK

**Comment by expresswallet** [comment]
ACK

**Comment by BlockScanUK** [comment]
ACK

**Comment by HashDen** [comment]
ACK

**Comment by TheBlueMatt** [comment]
LOL

**Comment by SREQ** [comment]
ACK

**Comment by dicecoins** [comment]
ACK

**Comment by jameshilliard** [comment]
@hellobitcoins @ydtr @expresswallet @BlockScanUK @HashDen @SREQ @dicecoins what were the results of your [testing](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review) did you find any issues?

**Comment by BitcoinPros** [comment]
ACK

My "tests" showed that blocks are full and we are in danger of a Bitcoin collapse.

**Comment by jameshilliard** [comment]
@jgarzik I think this increase is too fast if SegWit is deployed with it since SegWit basically acts as a multiplier.

**Comment by BitcoinPros** [comment]
@jameshilliard SegWit destroys the security of Bitcoin. Jeff already rejected it.

**Comment by jameshilliard** [comment]
@BitcoinPros I think you are looking for "Concept ACK" since "ACK" implies actually testing a pull request. What about SegWit destroys the security of Bitcoin?

**Comment by arsenische** [comment]
Block size should be constrained ONLY by technology that grows exponentially:

block_size_limit = base_bock_size_limit * projected_growth_rate ^ n_of_blocks_since_the_fork

The only question is what are the good enough values for base_block_size_limit and the projected_growth_rate.

Isn't it obvious that both should be greater than 1?

**Comment by jameshilliard** [comment]
@arsenische 

> But block size should be constrained ONLY by technology that grows exponentially:

The main problem is that there are a lot of bottlenecks that can be both hard to predict. Any sort of exponential or even linear scaling can be risky if the rate is too high or if they don't have a ceiling. I prefer to be on the safe side and not schedule any increases that we aren't fairly confident we can handle in the near term and I would prefer to not have any long term scheduled increase as the future is hard to predict. I think for this proposal we need to factor in the SegWit multiplier and hard cap it at 8MB or less.

**Comment by arsenische** [comment]
Isn't gradual exponential increase is safe enough? If it becomes obvious that there is a high risk of running into problems in 6 months or so then it would be easy to soft fork and decrease the chosen constants.

**Comment by jameshilliard** [comment]
@arsenische no far from it IMO, I don't think even a linear increase is safe if it is schedule to go over what we can currently handle or are reasonably confident we can handle in the near future, anyways this PR is obviously being brigaded from reddit so I'm going to stop commenting here, if you have any more questions feel free to contact me directly.

**Comment by laisee** [comment]
Concept ACK

**Comment by DavidVorick** [comment]
@arsenische It's not always clearly in the miner's best interests to reduce a block size. I think today's miners would definitely go along with a soft fork to reduce the block size if they felt it was necessary, but nobody wants to reduce the throughput of Bitcoin, especially large miners who benefit from larger block sizes (because it makes it more difficult for small miners to compete).

You may need a hardfork to reduce the blocksize (which means that it won't matter if the miners upgrade or not), but that would require getting all of the exchanges and merchants on board... seems pretty unlikely. Politically speaking, it seems that block size increase is a one way journey, especially because I don't think the public would recognize a 'blocks too big' crisis even mining pools stopped existing and all that was left were a small handful of datacenter miners.

I also don't think that setting the blocksize to grow exponentially will solve the political arguments or permanently put the blocksize debate to rest. If you set the exponent too high, and blocksize outpaces technologies ability to keep up, you end up in a disaster where you are politically manoeuvring to have the blocksize decreased. If you set the exponent  too small, there will be obvious gaps between the capabilities of technology and the throughput of the Bitcoin network, and we'll be right back to where we started with the ongoing debate.

That said, NACK.

The segregated witness proposal I feel is a much stronger solution, is on track for quick deployment, is a soft fork that preserves compatibility with old nodes (old nodes will be able to send money, receive money, and even receive money from segwit addresses - a point I think is lost on many), and raises the effective blocksize by *nearly* as much.

**Comment by arsenische** [comment]
@petertodd: that's correct that we may be forced to decrease it, it is easy to do with a soft fork if needed.

> Any increase must come with stakeholders signing onto the concept that this is a likely one-time increase. If not theres no reason for us to release code with an increase - let a competing team do that.

Quite the opposite. Artificial limitation to an arbitrary fixed limit should be signed by stakeholders. Currently they are [signing in favour of exponential increase](http://bitcoinocracy.com/arguments/bip101-is-better-than-the-status-quo).

**Comment by RussianNeuroMancer** [comment]
https://blockchain.info/charts/n-transactions-excluding-popular?timespan=2year&showDataPoints=true&daysAverageString=1&show_header=true&scale=0

Now we have ~2x every year. This proposal simply doesn't catch up with transaction growth - we will start to hit limits very soon if this get landed.

> it is not intended to handle exponential growth

Then I don't see the point.

**Comment by FreeMenInc** [comment]
@jameshilliard BitcoinPros is right. Soft forks kill security https://medium.com/@octskyward/on-consensus-and-forks-c6a050c792e7#.kbz5yvlrk and in SegWitness itself kills security: "splitting the signatures away from transactions, and creating a new apparently-unsigned version of the transaction opens the possibility of some network attack" https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2015-December/011976.html and does NOT improve scalability.

Bitcoin is breaking up as we speak. BLOCKS ARE FULL. If the size is not lifted RIGHT NOW a billion dollars of investment will dry up when BITCOIN COMPANIES ALL FAIL BECAUSE THEY WERE UNABLE TO DEPLOY NEW APPLICATIONS ON THE BLOCKCHAIN and BITCOIN WILL BE OVER. This is not a theory IT IS ALREADY HAPPENING. This is why the price fell from over $1000!!!!

"I have never seen this much disconnect between user wishes and dev outcomes in 20+ years of open source", @jameshilliard you SHITTY DEVELOPERS need to know your place: BEHIND THE KEYBOARD FOLLOWING ORDERS FROM REAL BUSINESS LEADERS. Your CONSTANT NEGATIVITY is KILLING BITCOIN.

You don't get to decide this. Your job now is to test it and carry out the decision.

Is this update enough? NO. But it is a promise to provide increases WHEN NEEDED. Without getting this in the next few days BITCOIN WILL DIE.

ACK ACK ACK FUCKING ACK

**Comment by DarlingStumpy** [comment]
ACK  This should be merged right now. The support is universal and any bug-fixes or improvements can be worked out prior to release. The most important thing is that someone take charge.

**Comment by SugarBub** [comment]
ACK. Why is this not merged already?

**Comment by jgarzik** [comment]
PR locked and vote brigading removed.

**Comment by BlockX2week** [comment]
Предлагаю : Размер блока  максимальный  = пересчет каждые 2016 блоков (2 недели) среднего размера блока умноженного на 2 (Max size block((2016*(<~1MB))/2016)*2)

**Comment by arowser** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**